components/ephemeral_storage/browser/ephemeral_storage_browsertest.cc

```diff
@@ -0,0 +1,164 @@
+
+class EphemeralStorageBrowserTest : public BraveNetworkDelegateBrowserTest {
+ public:
+  EphemeralStorageBrowserTest() : BraveNetworkDelegateBrowserTest() {}
+
+  void SetUpOnMainThread() override {
+    BraveNetworkDelegateBrowserTest::SetUpOnMainThread();
+    ephemeral_storage_url_ =
+        https_server_.GetURL("a.com", "/ephemeral_storage.html");
+    simple_url_ = https_server_.GetURL("a.com", "/simple.html");
+  }
+
+ protected:
+  GURL ephemeral_storage_url_;
+  GURL simple_url_;
+
+ private:
+  DISALLOW_COPY_AND_ASSIGN(EphemeralStorageBrowserTest);
+};
+
+IN_PROC_BROWSER_TEST_F(EphemeralStorageBrowserTest,
+                       StorageClearedOnMainFrameLoad) {
+  auto storage_types =
+      std::vector<StorageType>{StorageType::Session, StorageType::Local};
+  for (auto storage_type : storage_types) {
+    ui_test_utils::NavigateToURL(browser(), ephemeral_storage_url_);
+
+    auto* web_contents = browser()->tab_strip_model()->GetActiveWebContents();
+    content::RenderFrameHost* iframe =
+        content::ChildFrameAt(web_contents->GetMainFrame(), 0);
+    content::RenderFrameHost* third_party_iframe_a =
+        content::ChildFrameAt(web_contents->GetMainFrame(), 1);
+    content::RenderFrameHost* third_party_iframe_b =
+        content::ChildFrameAt(web_contents->GetMainFrame(), 2);
+
+    SetStorageValueInFrame(web_contents->GetMainFrame(), "key", "main",
+                           storage_type);
+    SetStorageValueInFrame(third_party_iframe_a, "key", "thirdparty",
+                           storage_type);
+
+    EXPECT_EQ("main", GetStorageValueInFrame(web_contents->GetMainFrame(),
+                                             "key", storage_type));
+    EXPECT_EQ("main", GetStorageValueInFrame(iframe, "key", storage_type));
+    EXPECT_EQ("thirdparty", GetStorageValueInFrame(third_party_iframe_a, "key",
+                                                   storage_type));
+    EXPECT_EQ("thirdparty", GetStorageValueInFrame(third_party_iframe_b, "key",
+                                                   storage_type));
+
+    ui_test_utils::NavigateToURL(browser(), ephemeral_storage_url_);
+
+    web_contents = browser()->tab_strip_model()->GetActiveWebContents();
+    iframe = content::ChildFrameAt(web_contents->GetMainFrame(), 0);
+    third_party_iframe_a =
+        content::ChildFrameAt(web_contents->GetMainFrame(), 1);
+    third_party_iframe_b =
+        content::ChildFrameAt(web_contents->GetMainFrame(), 2);
+
+    EXPECT_EQ("main", GetStorageValueInFrame(web_contents->GetMainFrame(),
+                                             "key", storage_type));
+    EXPECT_EQ("main", GetStorageValueInFrame(iframe, "key", storage_type));
+    EXPECT_EQ(nullptr, GetStorageValueInFrame(third_party_iframe_a, "key",
+                                              storage_type));
+    EXPECT_EQ(nullptr, GetStorageValueInFrame(third_party_iframe_b, "key",
+                                              storage_type));
+  }
+}
+
```

Comment:
we need tests that show proper separation of storage between domains to verify this functionality